ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DeprecatedStatementsChecker.java

```diff
@@ -0,0 +1,122 @@
+import io.confluent.ksql.parser.tree.AstNode;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.Join;
+import io.confluent.ksql.parser.tree.JoinedSource;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Relation;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.WithinExpression;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+public class DeprecatedStatementsChecker {
+  public enum Deprecations {
+    DEPRECATED_STREAM_STREAM_OUTER_JOIN_WITH_NO_GRACE(
+        "DEPRECATION NOTICE: Left/Outer stream-stream joins statements without a GRACE PERIOD "
+            + "will not be accepted in a future ksqlDB version.\n"
+            + "Please use the GRACE PERIOD clause as specified in "
+            + "https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/"
+            + "select-push-query/"
+    );
+
+    private final String noticeMessage;
+
+    Deprecations(final String noticeMessage) {
+      this.noticeMessage = noticeMessage;
+    }
+
+    public String getNoticeMessage() {
+      return noticeMessage;
+    }
+  }
+
+  final MetaStore metaStore;
+
+  DeprecatedStatementsChecker(final MetaStore metaStore) {
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+  }
+
+  public Optional<Deprecations> checkStatement(final Statement statement) {
+    if (isLeftOrOuterStreamStreamJoinWithoutGraceStatement(statement)) {
+      return Optional.of(Deprecations.DEPRECATED_STREAM_STREAM_OUTER_JOIN_WITH_NO_GRACE);
+    }
+
+    return Optional.empty();
+  }
+
+  private boolean isLeftOrOuterStreamStreamJoinWithoutGraceStatement(final Statement statement) {
+    if (statement instanceof CreateAsSelect) {
+      return isLeftOrOuterStreamStreamJoinWithoutGraceQuery(
+          ((CreateAsSelect) statement).getQuery());
+    } else if (statement instanceof Query) {
+      return isLeftOrOuterStreamStreamJoinWithoutGraceQuery((Query) statement);
+    }
+
+    return false;
+  }
+
+  private boolean isLeftOrOuterStreamStreamJoinWithoutGraceQuery(final Query query) {
+    if (query.getFrom() instanceof Join) {
+      final Join join = (Join) query.getFrom();
+
+      // check left joined source is a stream
+      if (!isStream(join.getLeft())) {
+        return false;
+      }
+
+      for (final JoinedSource joinedSource : join.getRights()) {
+        final JoinedSource.Type joinType = joinedSource.getType();
+        if (joinType == JoinedSource.Type.LEFT || joinType == JoinedSource.Type.OUTER) {
+          // check right joined source is a stream
+          if (!isStream(joinedSource.getRelation())) {
+            break;
+          }
+
+          if (!joinedSource.getWithinExpression().flatMap(WithinExpression::getGrace).isPresent()) {
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private boolean isStream(final Relation relation) {
```

Comment:
Not sure if I can follow?

If the left join node is a stream-table join, one source would be a table, but the result would still be a stream?